Move functorch decompositions to PyTorch

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76311

This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Move functorch decompositions to PyTorch

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

zou3519 commented on test/test_ops.py:
```diff
@@ -231,6 +238,36 @@ def test_reference_testing(self, device, dtype, op):
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(_decomp_test_ops, allowed_dtypes=(torch.float64, torch.long, torch.complex128))
+    def test_decomp(self, device, dtype, op):
+        hit = False
+
+        class DecompMode(torch.Tensor):  # TODO: don't subclass from tensor
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args, kwargs):
+                # TODO: this no_dispatch call shouldn't be necessary after
+                # https://github.com/pytorch/pytorch/pull/75965
+                with no_dispatch():
+                    if func in decomposition_table:
+                        nonlocal hit
+                        hit = True
+                        return decomposition_table[func](*args, **kwargs)
+                    else:
+                        return func(*args, **kwargs)
+
```
@Chillee wrote tests for decompositions over at https://github.com/pytorch/functorch/blob/955decd01aa7fc391f2bfbbb7ae5de0f9ed8ed0a/test/test_ops.py#L1349, should we just upstream those? These cover autograd testing and some additional dtypes.

zou3519 commented on torch/_decomp/__init__.py:
```diff
@@ -0,0 +1,76 @@
+            registry = decomposition_table
+
+        def add_op_to_table(aten_op):
+            # Converts aten.foo to aten.foo.default
+            # Done so I can be lazy and not write default on all of these ops
+            if not isinstance(aten_op, torch._ops.OpOverload):
```
nit: Could this be confusing? I'm worried someone can interpret aten.foo as "this is the decomposition for all overloads of aten.foo"

ezyang replied:
I'm OK with changing this, I was mostly trying to keep in line with the old code as much as possible. @Chillee any thoughts?

Chillee replied:
hmmm.... I think it's potentially fine.

I wouldn't be massively opposed to forcing an explicit use of `default` on every decomposition - just makes things a bit uglier.

ezyang replied:
I feel there is a more fundamental API design question we should resolve one way or another, which is whether or not it's OK to say torch.ops.aten.foo when we mean torch.ops.aten.foo.default. I know @anjali411 hates it. This area generally seems like an extreme attractive nuisance in the API right now. If we're going to keep the conversion, we should commit to "whenever I get an OpOverloadPacket where I expect an OpOverload, I will grab the default overload" and apply it consistently in all our torch.ops handling APIs.

anjali411 replied:
> whether or not it's OK to say torch.ops.aten.foo when we mean torch.ops.aten.foo.default

I don't think it's a good idea. Not every operator is guaranteed to have a default overload. And in many cases like add, default overload is probably not what you want since it corresponds to the following op: `aten::add(Scalar a, Scalar b) -> (Scalar)`

On a more high level, it seems bad to infer the packet as the default overload by default because calling those two operators on the same set of arguments can lead to different result (because of jit overload resolution). 

Does it make sense to register decompositions for the packets instead of the overloads? To put it another way, do we plan to register different decompositions for different overloads?

ezyang replied:
Hypothetically, you could register a decomp for a packet, in which case you are saying "I will do overload resolution myself", but for the most part I expect people to register decomps for specific overloads. It's quite a bit nicer.

zou3519 reviewed (commented):
Lint failures look legit. I want to know what the testing plan is: there is some decomposition testing in functorch that handles autograd and other devices that could be upstreamed here (is that coming later?). Is this a candidate for cross ref testing?

ezyang commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -629,6 +629,7 @@ def __init__(self,
                                                    # defaults to the value of `supports_forward_ad`
                  check_inplace_batched_forward_grad=None,   # whether to check batched forward grad when doing gradcheck
                                                             # defaults to the value of `check_batched_forward_grad`
+                 check_decomp=False,  # whether to check against torch._decomp
```
My main goal is I want to be able to run tests for all decompositions quickly; I don't want to waste time churning through forward-backwards for everything in the OpInfo, when only twenty of the operators actually do anything nontrivial.

The test cannot easily query if a decomposition exists for an operator because OpInfos deal with torch API, and decompositions are registered against ATen API. I could dump a table mapping torch functions to aten functions (this would be relatively easy to do), but you now have the same problem of keeping the table up to date (though, I suppose, one could have a separate test testing that the table is up to date). WDYT?

mruberry replied:
Yeah that all sounds reasonable. A table is interesting but I wouldn't want to make it a requirement today but more like an option later. But don't all the OpInfos specify their ATen names anyway? Are those not usable?

ezyang replied:
Ah, in this case it's because I need to know what the ATen names of the backwards are, but I suppose we can add that instead!

ezyang committed:
Update on "Move functorch decompositions to PyTorch"


This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch.

I also added some OpInfo testing and fix some complex dtype related bugs (though not all of them); search for TODO in the decompositions to see what I didn't fix. OpInfo tests only cover operators which are directly covered via OpInfos, so this excludes CUDA only operators as well as backward only operators (which have no OpInfo coverage).

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee reviewed (approved):
LGTM!

ezyang committed:
Update on "Move functorch decompositions to PyTorch"


This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee commented on test/test_decomp.py:
```diff
@@ -0,0 +1,458 @@
+        "cpu",
+        torch.bfloat16,
+        "nn.functional.layer_norm",
+    ),  # "batch_norm" not implemented for 'BFloat16'
```
Why is this needed - is it not already covered by OpInfo dtype support?

ezyang replied:
I am not sure but empirically the test failed even after I made us respect backward dtypes.

ezyang commented:
@zou3519 latest update now incorporates functorch's tests plus improvements. True crossref testing I think will late for later.

ezyang committed:
Update on "Move functorch decompositions to PyTorch"


This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move functorch decompositions to PyTorch"


This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move functorch decompositions to PyTorch"


This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

mruberry commented on test/test_decomp.py:
```diff
@@ -0,0 +1,481 @@
+        self.do_cross_ref(device, dtype, op, run_all=True)
+
+    def do_cross_ref(self, device, dtype, op, *, run_all):
+        if (torch.device(device).type, dtype, op.name) in CROSS_REF_EXCLUDE_SET or (
```
fyi maybe these skips could be modeled as skips on the OpInfos themselves for consistency?

mruberry replied:
I guess the issue with that is you'd have to skip multiple tests, which is kinda a pain?

ezyang replied:
I thought about it... it just seemed a lot harder to handle the skips if they were smeared out through the OpInfo database, rather than centralized on the test itself. I guess this is a general point on OpInfo skips: I think they should be attached to the relevant tests, rather than the OpInfo entries themselves.

ezyang committed:
Update on "Move functorch decompositions to PyTorch"


This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this

ezyang commented:
closed by https://github.com/pytorch/pytorch/commit/a3f10ec28176b8bda21e54ace5c0fae71868f509

zou3519 commented:
> @zou3519 latest update now incorporates functorch's tests plus improvements. True crossref testing I think will late for later.

Awesome, thanks for incorporating those tests!

byronyi commented on torch/_decomp/decompositions.py:
```diff
@@ -0,0 +1,795 @@
+@cast_for_opmath
+def native_dropout_decomposition(input, p, generator=None):
+    bool_mask = torch.rand_like(input) < p
+    res = bool_mask * input * float(1.0 / p)
```
I think these two lines should use (1-p)? cc @ezyang @zou3519 @Chillee @anjali411
